File: apps/api/src/routes/reports.ts
import { Router } from "express";
import type { Request } from "express";
import { authMiddleware } from "../middleware/auth.js";
import { requirePermission } from "../middleware/permissions.js";
import {
  getAvailabilityReport,
  getExecutiveReport,
  getProblemsReport,
  getProjectsReport,
  parsePeriod,
} from "../services/report-service.js";
import { generatePdf } from "../services/pdf-report-generator.js";

const router = Router();

router.use(authMiddleware);

const getQueryStringValue = (
  value: unknown,
): string | undefined =>
  typeof value === "string" && value.length > 0 ? value : undefined;

router.get("/availability", requirePermission("reports", "view"), async (req: Request, res) => {
  try {
    const period = {
      start: getQueryStringValue(req.query.start),
      end: getQueryStringValue(req.query.end),
    };
    const data = await getAvailabilityReport(
      parsePeriod(period),
    );
    res.json({ success: true, data });
  } catch (err) {
    console.error("❌ [REPORTS/AVAILABILITY]:", err);
    res.status(500).json({ success: false, error: "Erro interno" });
  }
});

// POST /api/reports/availability/pdf — exporta relatório de disponibilidade em PDF
router.post("/availability/pdf", requirePermission("reports", "view"), async (req: Request, res) => {
  try {
    const period = {
      start: getQueryStringValue(req.body?.start),
      end: getQueryStringValue(req.body?.end),
    };
    const parsed = parsePeriod(period);
    const data = await getAvailabilityReport(parsed);
    const pdf = await generatePdf({
      type: "availability",
      data: data as unknown as Record<string, unknown>,
      periodLabel: `${parsed.start} a ${parsed.end}`,
    });
    console.log(`✅ [REPORTS/AVAILABILITY/PDF] Gerado: ${pdf.filename}`);
    res.json({ success: true, data: { url: pdf.url, filename: pdf.filename } });
  } catch (err) {
    console.error("❌ [REPORTS/AVAILABILITY/PDF]:", err);
    res.status(500).json({ success: false, error: "Erro ao gerar PDF" });
  }
});

router.get("/problems", requirePermission("reports", "view"), async (req: Request, res) => {
  try {
    const { start, end } = parsePeriod({
      start: getQueryStringValue(req.query.start),
      end: getQueryStringValue(req.query.end),
    });
    const productId = getQueryStringValue(req.query.productId);
    const problemCategory = getQueryStringValue(req.query.problemCategory);
    const data = await getProblemsReport({ start, end }, productId, problemCategory);
    res.json({ success: true, data });
  } catch (err) {
    console.error("❌ [REPORTS/PROBLEMS]:", err);
    res.status(500).json({ success: false, error: "Erro interno" });
  }
});

// POST /api/reports/problems/pdf — exporta relatório de problemas em PDF
router.post("/problems/pdf", requirePermission("reports", "view"), async (req: Request, res) => {
  try {
    const { start, end } = parsePeriod({
      start: getQueryStringValue(req.body?.start),
      end: getQueryStringValue(req.body?.end),
    });
    const productId = getQueryStringValue(req.body?.productId);
    const problemCategory = getQueryStringValue(req.body?.problemCategory);
    const data = await getProblemsReport({ start, end }, productId, problemCategory);
    const pdf = await generatePdf({
      type: "problems",
      data: data as unknown as Record<string, unknown>,
      periodLabel: `${start} a ${end}`,
    });
    console.log(`✅ [REPORTS/PROBLEMS/PDF] Gerado: ${pdf.filename}`);
    res.json({ success: true, data: { url: pdf.url, filename: pdf.filename } });
  } catch (err) {
    console.error("❌ [REPORTS/PROBLEMS/PDF]:", err);
    res.status(500).json({ success: false, error: "Erro ao gerar PDF" });
  }
});

router.get("/executive", requirePermission("reports", "view"), async (req: Request, res) => {
  try {
    const { start, end } = parsePeriod({
      start: getQueryStringValue(req.query.start),
      end: getQueryStringValue(req.query.end),
    });
    const productId = getQueryStringValue(req.query.productId);
    const groupId = getQueryStringValue(req.query.groupId);
    const data = await getExecutiveReport({ start, end }, productId, groupId);
    res.json({ success: true, data });
  } catch (err) {
    console.error("❌ [REPORTS/EXECUTIVE]:", err);
    res.status(500).json({ success: false, error: "Erro interno" });
  }
});

// POST /api/reports/executive/pdf — exporta relatório executivo em PDF
router.post("/executive/pdf", requirePermission("reports", "view"), async (req: Request, res) => {
  try {
    const { start, end } = parsePeriod({
      start: getQueryStringValue(req.body?.start),
      end: getQueryStringValue(req.body?.end),
    });
    const productId = getQueryStringValue(req.body?.productId);
    const groupId = getQueryStringValue(req.body?.groupId);
    const data = await getExecutiveReport({ start, end }, productId, groupId);
    const pdf = await generatePdf({
      type: "executive",
      data: data as unknown as Record<string, unknown>,
      periodLabel: `${start} a ${end}`,
    });
    console.log(`✅ [REPORTS/EXECUTIVE/PDF] Gerado: ${pdf.filename}`);
    res.json({ success: true, data: { url: pdf.url, filename: pdf.filename } });
  } catch (err) {
    console.error("❌ [REPORTS/EXECUTIVE/PDF]:", err);
    res.status(500).json({ success: false, error: "Erro ao gerar PDF" });
  }
});

router.get("/projects", requirePermission("reports", "view"), async (req: Request, res) => {
  try {
    const period = {
      start: getQueryStringValue(req.query.start),
      end: getQueryStringValue(req.query.end),
    };
    const data = await getProjectsReport(
      parsePeriod(period),
    );
    res.json({ success: true, data });
  } catch (err) {
    console.error("❌ [REPORTS/PROJECTS]:", err);
    res.status(500).json({ success: false, error: "Erro interno" });
  }
});

// POST /api/reports/projects/pdf — exporta relatório de projetos em PDF
router.post("/projects/pdf", requirePermission("reports", "view"), async (req: Request, res) => {
  try {
    const period = {
      start: getQueryStringValue(req.body?.start),
      end: getQueryStringValue(req.body?.end),
    };
    const parsed = parsePeriod(period);
    const data = await getProjectsReport(parsed);
    const pdf = await generatePdf({
      type: "projects",
      data: data as unknown as Record<string, unknown>,
      periodLabel: `${parsed.start} a ${parsed.end}`,
    });
    console.log(`✅ [REPORTS/PROJECTS/PDF] Gerado: ${pdf.filename}`);
    res.json({ success: true, data: { url: pdf.url, filename: pdf.filename } });
  } catch (err) {
    console.error("❌ [REPORTS/PROJECTS/PDF]:", err);
    res.status(500).json({ success: false, error: "Erro ao gerar PDF" });
  }
});

/**
 * GET /api/reports/files — lista relatórios PDF gerados anteriormente
 */
router.get("/files", requirePermission("reports", "view"), async (_req: Request, res) => {
  try {
    const { listUploadFiles } = await import("../infra/uploads.js");
    const files = await listUploadFiles("reports");
    // Filtra apenas PDFs
    const pdfs = files.filter((f) => f.filename.endsWith(".pdf"));
    res.json({ success: true, data: pdfs });
  } catch (err) {
    console.error("❌ [REPORTS/FILES]:", err);
    res.status(500).json({ success: false, error: "Erro ao listar relatórios" });
  }
});

export default router;

File: apps/backend/src/silo/api/routers/reports.py
from __future__ import annotations

from collections.abc import Mapping

from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse
from sqlalchemy.engine import Connection

from silo.api.dependencies import get_snapshot_db, require_permission
from silo.api.responses import build_success_payload, json_error_response
from silo.services.report_portal import (
    PdfArtifactTooLargeError,
    UnsupportedReportFilterError,
    generate_pdf,
    get_availability_report,
    get_availability_report_meta,
    get_executive_report,
    get_executive_report_meta,
    get_problems_report,
    get_problems_report_meta,
    get_projects_report,
    get_projects_report_meta,
    list_report_files,
    parse_period,
)

router = APIRouter(
    prefix="/api/reports",
    tags=["reports"],
    dependencies=[Depends(require_permission("reports", "view"))],
)


@router.get("/availability")
async def availability_report(request: Request, db: Connection = Depends(get_snapshot_db)):
    period = _parse_period_or_error(request.query_params)
    if isinstance(period, JSONResponse):
        return period

    try:
        data = get_availability_report(db, period)
        return build_success_payload(data, meta=get_availability_report_meta(period))
    except Exception:
        return json_error_response(500, "Erro interno")


@router.post("/availability/pdf")
async def availability_pdf(request: Request, db: Connection = Depends(get_snapshot_db)):
    body = await _request_json_object(request)
    period = _parse_period_or_error(body)
    if isinstance(period, JSONResponse):
        return period

    try:
        data = get_availability_report(db, period)
        pdf = generate_pdf(report_type="availability", data=data, period_label=f"{period['start']} a {period['end']}")
        return build_success_payload({"url": pdf["url"], "filename": pdf["filename"]})
    except PdfArtifactTooLargeError as exc:
        return json_error_response(413, "ARTIFACT_TOO_LARGE", data=exc.as_payload())
    except Exception:
        return json_error_response(500, "Erro ao gerar PDF")


@router.get("/problems")
async def problems_report(request: Request, db: Connection = Depends(get_snapshot_db)):
    period = _parse_period_or_error(request.query_params)
    if isinstance(period, JSONResponse):
        return period

    product_id = _optional_text(request.query_params.get("productId"))
    problem_category = _optional_text(request.query_params.get("problemCategory")) or _optional_text(request.query_params.get("problem_category"))

    try:
        data = get_problems_report(db, period, product_id, problem_category)
        return build_success_payload(data, meta=get_problems_report_meta(period))
    except Exception:
        return json_error_response(500, "Erro interno")


@router.post("/problems/pdf")
async def problems_pdf(request: Request, db: Connection = Depends(get_snapshot_db)):
    body = await _request_json_object(request)
    period = _parse_period_or_error(body)
    if isinstance(period, JSONResponse):
        return period

    product_id = _optional_text(body.get("productId"))
    problem_category = _optional_text(body.get("problemCategory")) or _optional_text(body.get("problem_category"))

    try:
        data = get_problems_report(db, period, product_id, problem_category)
        pdf = generate_pdf(report_type="problems", data=data, period_label=f"{period['start']} a {period['end']}")
        return build_success_payload({"url": pdf["url"], "filename": pdf["filename"]})
    except PdfArtifactTooLargeError as exc:
        return json_error_response(413, "ARTIFACT_TOO_LARGE", data=exc.as_payload())
    except Exception:
        return json_error_response(500, "Erro ao gerar PDF")


@router.get("/executive")
async def executive_report(request: Request, db: Connection = Depends(get_snapshot_db)):
    period = _parse_period_or_error(request.query_params)
    if isinstance(period, JSONResponse):
        return period

    product_id = _optional_text(request.query_params.get("productId"))
    group_id = _optional_text(request.query_params.get("groupId"))

    try:
        data = get_executive_report(db, period, product_id, group_id)
        return build_success_payload(data, meta=get_executive_report_meta(period))
    except UnsupportedReportFilterError:
        return json_error_response(400, "UNSUPPORTED_FILTER")
    except Exception:
        return json_error_response(500, "Erro interno")


@router.post("/executive/pdf")
async def executive_pdf(request: Request, db: Connection = Depends(get_snapshot_db)):
    body = await _request_json_object(request)
    period = _parse_period_or_error(body)
    if isinstance(period, JSONResponse):
        return period

    product_id = _optional_text(body.get("productId"))
    group_id = _optional_text(body.get("groupId"))

    try:
        data = get_executive_report(db, period, product_id, group_id)
        pdf = generate_pdf(report_type="executive", data=data, period_label=f"{period['start']} a {period['end']}")
        return build_success_payload({"url": pdf["url"], "filename": pdf["filename"]})
    except PdfArtifactTooLargeError as exc:
        return json_error_response(413, "ARTIFACT_TOO_LARGE", data=exc.as_payload())
    except UnsupportedReportFilterError:
        return json_error_response(400, "UNSUPPORTED_FILTER")
    except Exception:
        return json_error_response(500, "Erro ao gerar PDF")


@router.get("/projects")
async def projects_report(request: Request, db: Connection = Depends(get_snapshot_db)):
    period = _parse_period_or_error(request.query_params)
    if isinstance(period, JSONResponse):
        return period

    try:
        data = get_projects_report(db, period)
        return build_success_payload(data, meta=get_projects_report_meta(period))
    except Exception:
        return json_error_response(500, "Erro interno")


@router.post("/projects/pdf")
async def projects_pdf(request: Request, db: Connection = Depends(get_snapshot_db)):
    body = await _request_json_object(request)
    period = _parse_period_or_error(body)
    if isinstance(period, JSONResponse):
        return period

    try:
        data = get_projects_report(db, period)
        pdf = generate_pdf(report_type="projects", data=data, period_label=f"{period['start']} a {period['end']}")
        return build_success_payload({"url": pdf["url"], "filename": pdf["filename"]})
    except PdfArtifactTooLargeError as exc:
        return json_error_response(413, "ARTIFACT_TOO_LARGE", data=exc.as_payload())
    except Exception:
        return json_error_response(500, "Erro ao gerar PDF")


@router.get("/files")
async def report_files():
    try:
        return build_success_payload(list_report_files())
    except Exception:
        return json_error_response(500, "Erro ao listar relatórios")


async def _request_json_object(request: Request) -> dict[str, object]:
    try:
        body = await request.json()
    except Exception:
        return {}
    return body if isinstance(body, dict) else {}


def _parse_period_or_error(mapping: Mapping[str, object]) -> dict[str, str] | JSONResponse:
    try:
        return parse_period(dict(mapping))
    except Exception:
        return json_error_response(400, "Data inválida.")


def _optional_text(value: object | None) -> str | None:
    return value.strip() if isinstance(value, str) and value.strip() else None
